### PR TITLE
cadl,support emitter options

### PR DESCRIPTION
**Pull request by XiaofeiCao**
fix https://github.com/Azure/autorest.java/issues/1737

### CODE REVIEW

File: cadl-extension/src/main/java/com/azure/cadl/Main.java [hunk 1]
```diff
@@ -115,28 +92,58 @@ public static void main(String[] args) throws IOException {
...
-                LOGGER.error("Failed to format file: {}", outputFolderFinal + filePath, e);
+                LOGGER.error("Failed to format file: {}", filePath, e);
                 // but we continue so user can still check the file and see why format fails
             }
             formattedFiles.put(filePath, formattedSource);
         });
 
         // write output
         // java files
-        formattedFiles.forEach((filePath, formattedSource) -> cadlPlugin.writeFile(outputFolderFinal + filePath, formattedSource, null));
```
Comment (XiaofeiCao) [inline cadl-extension/src/main/java/com/azure/cadl/Main.java:126]: Align with autorest, resolve output-folder in `CadlPlugin`.

File: cadl-extension/src/emitter.ts [hunk 2]
```diff
@@ -7,7 +7,24 @@ import { CodeModelBuilder } from "./code-model-builder.js";
...
+
+export const $lib = createCadlLibrary({
+  name: "JavaEmitter",
+  diagnostics: {},
+  emitter: {
+      options: EmitterOptionsSchema,
+  },
+});
+
+export async function $onEmit(program: Program, options: EmitterOptions) {
```
Comment (XiaofeiCao) [inline cadl-extension/src/emitter.ts:27]: cadl emitter settings goes into `EmitterOptions`.

File: cadl-tests/cadl-project.yaml [hunk 3]
```diff
@@ -1,2 +1,4 @@
 emitters:
-  '@azure-tools/cadl-java': true
+  '@azure-tools/cadl-java':
+    generate-models: true
```
Comment (XiaofeiCao) [inline cadl-tests/cadl-project.yaml:3]: The settings will directly go into `CadlPlugin`'s `SETTINGS_MAP`, thus readable by `JavaSettings`.

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/CadlSettings.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.autorest.extension.base.plugin;
+
+public class CadlSettings {
```
Comment (weidongxu-microsoft) [inline extension-base/src/main/java/com/azure/autorest/extension/base/plugin/CadlSettings.java:6]: why do we have CadlSetting in shared module? Should it be in cadl-extension module?

File: cadl-tests/cadl-project.yaml (hunk 3 above)
Comment (weidongxu-microsoft) [reply cadl-tests/cadl-project.yaml:3]: we probably don't want to expose `generate-models` yet (not expect user to use it).

File: cadl-extension/src/test/java/com/azure/cadl/util/ModelUtilTests.java [hunk 5]
```diff
@@ -12,16 +12,19 @@
...
     // sadly ModelUtil.isGeneratingModel queries JavaSettings
     @BeforeAll
     public static void ensurePlugin() {
-        CadlPlugin plugin = new CadlPlugin(
-                new CadlPlugin.Options()
-                        .setNamespace("com.azure.client"));
+        CadlPlugin.Options options = new CadlPlugin.Options().withEmitterOptions(
+                new HashMap<String, Object>() {{
+                    this.put("namespace", "com.azure.client");
+                }});
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/test/java/com/azure/cadl/util/ModelUtilTests.java:26]: Don't use this `{{` in code. Use standard class.

File: cadl-extension/src/main/java/com/azure/cadl/Main.java [hunk 6]
```diff
@@ -115,28 +92,58 @@ public static void main(String[] args) throws IOException {
...
+                LOGGER.info("Read emitter options failed, emitter options json: {}", emitterOptionsJson);
+            }
+        }
+        if (!emitterOptions.containsKey("namespace")) {
+            if (codeModel.getLanguage().getJava() != null && !CoreUtils.isNullOrEmpty(codeModel.getLanguage().getJava().getNamespace())) {
+                emitterOptions.put("namespace", codeModel.getLanguage().getJava().getNamespace());
+            }
+        }
+        emitterOptions.put("output-folder", outputFolderFinal);
+        return new CadlPlugin.Options().withEmitterOptions(emitterOptions);
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/main/java/com/azure/cadl/Main.java:144]: Just `setEmitterOptions` or `addEmitterOptions`?

---

Why serialize to a `Map<String, Object>` and set the generic map via `withEmitterOptions`? We prefer a model whenever we know the data.
We write to and read from this "emitterOptions", then why keep ourselves in the dark about its possible content?

File: cadl-extension/src/emitter.ts [hunk 7]
```diff
@@ -7,7 +13,24 @@ import { CodeModelBuilder } from "./code-model-builder.js";
 import { dirname } from "path";
 import { fileURLToPath } from "url";
 
-export async function $onEmit(program: Program) {
+export interface EmitterOptions {}
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/emitter.ts:16]: Same here, we want exact type of possible options (like the code in tutorial). Not a blank `interface {}`.

**Comment by weidongxu-microsoft** [comment]
Use a model (instead of map) whenever you know/control the data.

File: cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java [hunk 8]
```diff
@@ -130,10 +123,6 @@ public String handlePartialUpdate(String filePath, String generatedContent) {
         SETTINGS_MAP.put("required-parameter-client-methods", true);
         SETTINGS_MAP.put("generic-response-type", true);
         SETTINGS_MAP.put("output-model-immutable", true);
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java:125]: If we want to change, these are the things we don't want. In Cadl, we would like to make the `JavaSettings` instance without this map.
However the singleton nature of the `JavaSettings` make it hard.

File: cadl-extension/src/main/java/com/azure/cadl/Main.java (hunk 6 above)
Comment (XiaofeiCao) [reply cadl-extension/src/main/java/com/azure/cadl/Main.java:144]: Yes, that's the part I felt weird about. My initial though was to ease the process of increasing options over time, but the map would confuse the dev about what the possible options could be. And if we want to control what options to accept, yes it's better to use a model.

File: extension-base/src/main/java/com/azure/autorest/extension/base/plugin/CadlSettings.java (hunk 4 above)
Comment (XiaofeiCao) [reply extension-base/src/main/java/com/azure/autorest/extension/base/plugin/CadlSettings.java:6]: I'm thinking about a global way of getting cadl settings like `output-folder`. Previously we had `JavaSettings.getAutorestSettings().getOutputFolder()`. I intended to add `JavaSettings.getCadlSettings().getOutputFolder()`. 

For now, the only place it is used is in `CadlPlugin`, so it seems too much to add a new class for it. Will inline the property in `CadlPlugin` instead.

File: cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java (hunk 8 above)
Comment (XiaofeiCao) [reply cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java:125]: I see. I would not change that for now.

File: cadl-extension/src/main/java/com/azure/cadl/Main.java [hunk 9]
```diff
@@ -115,28 +92,101 @@ public static void main(String[] args) throws IOException {
...
+            outputFolder = args[1];
+            if (!outputFolder.endsWith("/")) {
+                outputFolder += "/";
+            }
+        }
+
+        final String outputFolderFinal = outputFolder;
+
+        String emitterOptionsJson = Configuration.getGlobalConfiguration().get("emitterOptions");
+        Map<String, Object> emitterOptions = parseJsonAsMap(emitterOptionsJson);
```
[the remark this answers is not in the document]
Comment (weidongxu-microsoft) [reply cadl-extension/src/main/java/com/azure/cadl/Main.java:131]: I mean, would it be easier to just serialize it to a e.g. `EmitterOptions` model with `JsonProperties` annotation (than `getNonEmptyStringOrNull` `getBoolean` etc.)?

And maybe `CadlPlugin.Options` and be merged with it (we likely have `outputFolder` in emitter option)?

File: cadl-extension/src/emitter.ts [hunk 10]
```diff
@@ -7,7 +13,32 @@ import { CodeModelBuilder } from "./code-model-builder.js";
...
+}
+
+const EmitterOptionsSchema: JSONSchemaType<EmitterOptions> = {
+  type: "object",
+  additionalProperties: false,
+  properties: {
+    "namespace": { type: "string", nullable: true },
+    "service-name": { type: "string", nullable: true },
+    "partial-update": { type: "boolean", nullable: true },
+  },
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/emitter.ts:29]: we had high probability to add  `outputFolder` as well (I am not sure about the exact name though).

Also, should we use camelCase?

Comment (XiaofeiCao) [reply cadl-extension/src/emitter.ts:29]: Sure, will add that and change it to camel case.

Comment (weidongxu-microsoft) [reply cadl-extension/src/emitter.ts:29]: currently seems to be `outputPath` https://github.com/microsoft/cadl/issues/482

File: javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java [hunk 11]
```diff
@@ -317,7 +317,7 @@ protected void addSerializationImports(Set<String> imports, JavaSettings setting
     }
 
     /**
-     * Override parent setters if: 1. parent property does not have setter 2. child does not contain property
+     * Override parent setters if: 1. parent property has setter 2. child does not contain property
      * that shadow this parent property, otherwise overridden parent setter methods will collide with child setter
```
Comment (XiaofeiCao) [inline javagen/src/main/java/com/azure/autorest/template/ModelTemplate.java:321]: Not relevant, fix old doc.

File: cadl-extension/src/emitter.ts (hunk 10 above)
Comment (XiaofeiCao) [reply cadl-extension/src/emitter.ts]: Changed to outputPath.

File: cadl-extension/src/main/java/com/azure/cadl/Main.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (XiaofeiCao) [reply cadl-extension/src/main/java/com/azure/cadl/Main.java]: Changed. Added `outputPath` and `EmitterOptions`.

File: cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java [hunk 12]
```diff
@@ -189,4 +178,8 @@ public void message(Message message) {
                 break;
         }
     }
+
+    private String getOutputFolder() {
+        return (String) SETTINGS_MAP.get("output-folder");
```
Comment (weidongxu-microsoft) [inline cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java:183]: nit, seems easier to put the EmitterOptions in class var?

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
main should be fine with CI now.

File: cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java (hunk 12 above)
Comment (XiaofeiCao) [reply cadl-extension/src/main/java/com/azure/autorest/CadlPlugin.java:183]: sure

**Pull request merged**